zerver/data_import/microsoft_teams.py

```diff
@@ -0,0 +1,241 @@
+import logging
+import os
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import Any, Literal, TypeAlias
+from urllib.parse import SplitResult
+
+import requests
+from django.conf import settings
+from django.utils.timezone import now as timezone_now
+
+from zerver.data_import.import_util import (
+    ZerverFieldsT,
+    build_realm,
+    build_recipient,
+    build_subscription,
+    build_user_profile,
+    build_zerver_realm,
+    create_converted_data_files,
+    validate_user_emails_for_import,
+)
+from zerver.data_import.sequencer import NEXT_ID
+from zerver.data_import.slack import get_data_file
+from zerver.lib.export import do_common_export_processes
+from zerver.models.recipients import Recipient
+from zerver.models.users import UserProfile
+
+MicrosoftTeamsUserIdToZulipUserIdT: TypeAlias = dict[str, int]
+MicrosoftTeamsFieldsT: TypeAlias = dict[str, Any]
+
+
+@dataclass
+class MicrosoftTeamsUserRoleData:
+    global_administrator_user_ids: set[str]
+    guest_user_ids: set[str]
+
+
+@dataclass
+class ODataQueryParameter:
+    parameter: Literal["$filter", "$search", "$select"]
+    expression: str
+
+
+def get_microsoft_graph_api_data(
+    endpoint: str,
+    odata_parameters: list[ODataQueryParameter] | None = None,
+    **kwargs: Any,
+) -> Any:
+    if not kwargs.get("token"):
+        raise AssertionError("Microsoft authorization token missing in kwargs")
+    token = kwargs.pop("token")
+    api_url = f"https://graph.microsoft.com/v1.0{endpoint}"
+
+    parameters = {}
+    if odata_parameters is not None:
+        for parameter in odata_parameters:
+            assert parameter.parameter not in parameters
+            parameters[parameter.parameter] = parameter.expression
+
+    response = requests.get(
+        api_url, headers={"Authorization": f"Bearer {token}"}, params=parameters
+    )
+
+    if response.status_code != requests.codes.ok:
+        logging.info("HTTP error: %s, Response: %s", response.status_code, response.text)
+        raise Exception("HTTP error accessing the Microsoft Graph API.")
+    return response.json()
+
+
+def get_users_with_directory_role_id(
+    directory_role_id: str, api_token: str
+) -> list[MicrosoftTeamsFieldsT]:
+    result = get_microsoft_graph_api_data(
+        f"/directoryRoles/{directory_role_id}/members", token=api_token
+    )
+    return result["value"]
+
+
+def get_users_with_member_type(
+    member_type: Literal["Member", "Guest"], api_token: str
+) -> list[MicrosoftTeamsFieldsT]:
+    odata_parameter = [
+        ODataQueryParameter(parameter="$filter", expression=f"userType eq '{member_type}'"),
+    ]
+    result = get_microsoft_graph_api_data(
```

Comment:
Can you link to API documentation for the endpoints we're using? I guess the particular thing I'm wondering is whether we need pagination when there's a large amount of users? (Missing pagination was a sneaky bug we had in the Slack importer)